core/injector/application_injector.hpp

```diff
@@ -91,308 +92,347 @@ namespace kagome::injector {
       }
-      auto &cfg =
-          injector.template create<application::ConfigurationStorage &>();
+      ++index;
+    }
+    if (index >= authorities.size()) {
+      common::raise(InjectorError::INDEX_OUT_OF_BOUND);
+    }
+
+    initialized = std::make_shared<primitives::AuthorityIndex>(
+        primitives::AuthorityIndex{index});
+    return initialized.value();
+  }
 
-      initialized = std::make_shared<network::PeerList>(cfg.getBootNodes());
+  // extrinsic api listener getter
+  template <typename Injector>
+  sptr<api::Listener> get_extrinsic_api_listener(const Injector &injector) {
+    static auto initialized = boost::optional<sptr<api::Listener>>(boost::none);
+    // listener is used currently only for extrinsic api
+    // if other apis are required, need to
+    // implement lambda creating corresponding api service
+    // where listener is initialized manually
+    // in this case listener should be bound
+    if (initialized) {
       return initialized.value();
-    };
-
-    auto get_authority_index =
-        [initialized =
-             boost::optional<sptr<primitives::AuthorityIndex>>(boost::none)](
-            const auto &injector) mutable -> sptr<primitives::AuthorityIndex> {
-      if (initialized) {
-        return initialized.value();
-      }
+    }
+
+    auto &context = injector.template create<boost::asio::io_context &>();
+    auto &cfg = injector.template create<application::ConfigurationStorage &>();
+    auto extrinsic_tcp_version = boost::asio::ip::tcp::v4();
+    uint16_t extrinsic_api_port = cfg.getExtrinsicApiPort();
+    api::ListenerImpl::Configuration listener_config{
+        boost::asio::ip::tcp::endpoint{extrinsic_tcp_version,
+                                       extrinsic_api_port}};
+    auto &&http_session_config =
+        injector.template create<api::HttpSession::Configuration>();
+
+    initialized = std::make_shared<api::ListenerImpl>(
+        context, listener_config, http_session_config);
+    return initialized.value();
+  }
 
-      auto core = injector.template create<sptr<runtime::Core>>();
-      auto &keys = injector.template create<application::KeyStorage &>();
-      auto &&local_pair = keys.getLocalEd25519Keypair();
-      auto &public_key = local_pair.public_key;
-      auto &&authorities_res =
-          core->authorities(primitives::BlockId(primitives::BlockNumber{0}));
-      if (authorities_res.has_error()) {
-        common::raise(authorities_res.error());
-      }
-      auto &&authorities = authorities_res.value();
-
-      uint64_t index = 0;
-      for (auto &auth : authorities) {
-        if (public_key == auth.id) {
-          break;
-        }
-        ++index;
-      }
-      if (index >= authorities.size()) {
-        common::raise(InjectorError::INDEX_OUT_OF_BOUND);
-      }
+  // block storage getter
+  template <typename Injector>
+  sptr<blockchain::BlockStorage> get_block_storage(const Injector &injector) {
+    static auto initialized =
```

Comment:
I don't like how sharing is implemented. Previously it was one per injector, now it is one per application.